small I2C mux fixes

b0xcat opened the pull request:
Hi Skyler,

Here are the changes I needed to do to get the badge talking to i2c devices on the hexpansion ports. I was able to successfully talk to an i2c eeprom after configuring the gpio expanders to enable power to the hexpansion slots.

One small thing I haven't done yet is handling errors that might happen when switching the downstream port. I could take the cmd_set_downstream out of cmd_link_create to make it easier but I wasn't sure what would be the neatest way to do it.

b0xcat committed:
tildagon_i2c: Allow access to the system bus for now (needed to poke the gpio expanders)

b0xcat committed:
tildagon_i2c: Set TILDAGON_HOST_I2C_SDA and SCL to GPIO 45 and 46

b0xcat committed:
tca9548: Set downstream i2c mux bus in a separate i2c transaction, as adding additional commands to an i2c cmd link after stopping is not supported

b0xcat committed:
tildagon_i2c: Separate downstream switching from link creation and handle errors

b0xcat commented:
Did an attempt at error handling

Skyler84 commented:
i think putting the call to set the downstream port would be best in the cmd begin function and we can get rid of the create wrapper. i'll be on my laptop again shortly so i can do that and add the thread safety so there's no possibility of the port getting changed by a thread race condition

b0xcat commented:
Alright sounds good!

Skyler84 committed:
Revert "tca9548a: remove semaphores"

This reverts commit a09faffa3937950b4b3343e413b452ab44e440b1.

Re-added semaphores (but changed them to mutexes)
Restructured I2C wrapper slightly and removed the create_link function
as it's now just a plain wrapper with nothing special in it.

Skyler84 committed:
i2c: fix incorrect i2c mux port assignment

self->port in the mux obj is actually the upstream i2c port number, not
the downstream port, this has to be passed into the function instead!

Skyler84 committed:
i2c-mux: Add optimisation to i2c mux handling

only change downstream port when required.
Additionally don't expose the api to directly set the downstream, this
should be internal only. We don't want anyone changing the downstream
port without our knowledge! (it's also not thread safe on its own now)

Skyler84 committed:
i2c-mux: fix const bypass and mutex initialisation
There were a few bits that didn't build correctly.

Skyler84 committed:
i2c-mux: mutexes need not be given before being taken.

b0xcat committed:
Merge pull request #2 from Skyler84/skyler-i2c-c-mux-fix-2

i2c-mux: Add optimisation to i2c mux handling
